src/main/kotlin/net/djvk/fireflyPlaidConnector2/transactions/PlaidFireflyTransaction.kt

```diff
@@ -0,0 +1,127 @@
+package net.djvk.fireflyPlaidConnector2.transactions
+
+import net.djvk.fireflyPlaidConnector2.api.firefly.models.TransactionTypeProperty
+import net.djvk.fireflyPlaidConnector2.api.plaid.models.Transaction
+import java.time.OffsetDateTime
+import java.time.ZoneId
+
+/**
+ * Represents a single non-transfer transaction, including both the Plaid and the Firefly representation of it.
+ */
+sealed interface PlaidFireflyTransaction {
+    val amount: Double
+    val transactionId: String
+    val fireflyAccountId: Int
+    val fireflyTransaction: FireflyTransactionDto?
+    val plaidTransaction: Transaction?
+    fun getTimestamp(zoneId: ZoneId): OffsetDateTime
+
+    companion object {
+        /**
+         * Wraps Plaid and Firefly transactions into PlaidFireflyTransaction objects, joining them together when
+         * matching transactions are identified.
+         */
+        fun match(
+            plaidTxs: List<Transaction>,
+            fireflyTxs: List<FireflyTransactionDto>,
+            accountMap: Map<String, Int>
+        ): List<PlaidFireflyTransaction> {
+            val plaidById = plaidTxs.groupBy { FireflyTransactionExternalIdIndexer.getExternalId(it.transactionId) }
+            val fireflyByExtId = fireflyTxs.groupBy { it.tx.externalId }
+            val externalIds = plaidById.keys.union(fireflyByExtId.keys)
+
+            return externalIds.flatMap {
+                val matchingPlaid = plaidById[it] ?: listOf()
+                val matchingFirefly = fireflyByExtId[it] ?: listOf()
+
+                // For all transactions that do not have an external ID, or if we've found more matching transactions
+                // than we expected to find, return them without attempting to combine.
+                if (it == null || matchingFirefly.size > 1 || matchingPlaid.size > 1) {
+                    val convertedFirefly = matchingFirefly.map { FireflyTransaction(it) }
+                    val convertedPlaid = matchingPlaid.map {
```

Comment:
The `it` variables in these `map` calls are shadowing the `it` from the `flatMap` above; you should probably name the variables in the lambdas that are not inline.

Reply:
Thanks, still getting the hang of Kotlin. Fixed in https://github.com/nprzy/firefly-plaid-connector-2/commit/9713978338a6b6631f3e38df5e60e4f92578ae34